File: src/main/java/top/kmar/mi/api/electricity/EleTileEntity.java
package top.kmar.mi.api.electricity;

import net.minecraft.util.EnumFacing;
import net.minecraft.util.math.BlockPos;
import net.minecraftforge.common.capabilities.Capability;
import top.kmar.mi.api.araw.interfaces.AutoSave;
import top.kmar.mi.api.capabilities.ele.EleCapability;
import top.kmar.mi.api.capabilities.ele.IStorage;
import top.kmar.mi.api.electricity.clock.OverloadCounter;
import top.kmar.mi.api.electricity.info.EleEnergy;
import top.kmar.mi.api.tools.BaseTileEntity;

import javax.annotation.Nullable;
import java.util.HashSet;
import java.util.Set;

/**
 * 机器的父类，其中包含了机器的一些默认实现
 * @author EmptyDreams
 */
public abstract class EleTileEntity extends BaseTileEntity {
	
	/** 空的能量 */
	private static final EleEnergy EMPTY_ENERGY = new EleEnergy(0, EleEnergy.ZERO);
	
	/** 当前包含的能量 */
	@AutoSave private int nowEnergy;
	/** 可存储的最大能量 */
	private int maxEnergy = 0;
	/** 超载计时器 */
	private OverloadCounter counter;
	/** 存储已连接的方块 */
	@AutoSave
	private final Set<BlockPos> linkedBlocks = new HashSet<>(3);
	
	/**
	 * <p>当电器接收电能时调用，该方法只在服务端调用
	 * <p>该方法还用来处理当接受到超过电器可承受范围的电压时增加计数器的操作
	 * @param energy 这次接收的能量
	 * @return 返回false可以阻止此次电能的消耗
	 */
	public abstract boolean onReceive(EleEnergy energy);
	
	/**
	 * 当电器输出电能时调用，该方法只在服务端调用
	 * @param energy 这次输出的能量
	 * @return 返回false可以组织此次电能的输出
	 */
	public boolean onExtract(EleEnergy energy) {
		return true;
	}
	
	/**
	 * 是否可以从指定方向输入电能
	 * @param facing 相对当前方块的方向，为null表示任意方向
	 */
	public abstract boolean isReceiveAllowable(EnumFacing facing);
	
	/**
	 * 是否可以从指定方向输出电
	 * @param facing 相对当前方块的方向，为null表示任意方向
	 */
	public abstract boolean isExtractAllowable(EnumFacing facing);
	
	/**
	 * 获取输出电压
	 * @return 如果不能输出则返回 0
	 */
	public abstract int getExVoltage();
	
	/** 获取当前所需能量 */
	public int getDemandEnergy() {
		return getMaxEnergy() - getNowEnergy();
	}
	
	/** 是否可以连接指定方向的方 */
	public boolean canLinkEle(EnumFacing facing) {
		return hasCapability(EleCapability.ENERGY, facing);
	}
	/** 连接指定方块 */
	public boolean linkEle(BlockPos pos) {
		if (linkedBlocks.contains(pos)) return true;
		return linkedBlocks.add(pos);
	}
	/** 取消链接指定方块 */
	public boolean unLink(BlockPos pos) { return linkedBlocks.remove(pos); }
	
	/**
	 * 判断方块某个方向是否含有指定能力<br>
	 * 要求：{@link #getCapability(Capability, EnumFacing)}返回非null值时该方法必须返回true，
	 * 该方法返回true时{@link #getCapability(Capability, EnumFacing)}必须返回非null值
	 * @param capability 能力
	 * @param facing 方向
	 */
	@Override
	public boolean hasCapability(Capability<?> capability, @Nullable EnumFacing facing) {
		if (super.hasCapability(capability, facing)) return true;
		return capability == EleCapability.ENERGY;
	}
	
	/**
	 * 获取方块某个方向的能力，能力不存在时返回null
	 * @param capability 指定的能力
	 * @param facing 方向
	 * @param <T> 能力类型
	 * @see #hasCapability(Capability, EnumFacing) 
	 */
	@Nullable
	@Override
	public <T> T getCapability(Capability<T> capability, @Nullable EnumFacing facing) {
		if (capability == EleCapability.ENERGY) {
			return EleCapability.ENERGY.cast(storage);
		}
		return super.getCapability(capability, facing);
	}

	/** 获取现在的能量 */
	public int getNowEnergy() { return nowEnergy; }
	/** 设置现在的能量 */
	public void setNowEnergy(int nowEnergy) { this.nowEnergy = Math.min(Math.max(nowEnergy, 0), getMaxEnergy()); }
	/** 增加能量 */
	@SuppressWarnings("UnusedReturnValue")
	public int growEnergy(int grow) {
		int value = getNowEnergy() + grow;
		setNowEnergy(value);
		return value - getNowEnergy();
	}
	/**
	 * 减少能量
	 * @param shrink 要减少的数量
	 * @return 减少操作是否成功，若剩余能量小于shrink则返回false
	 */
	public boolean shrinkEnergy(int shrink) {
		int value = getNowEnergy() - shrink;
		if (value < 0) return false;
		setNowEnergy(value);
		return true;
	}
	/** 获取计数器 */
	public OverloadCounter getCounter() { return counter; }
	/** 设置计数器 */
	protected void setCounter(OverloadCounter counter) { this.counter = counter; }
	/** 获取可存储的能量值 */
	public int getMaxEnergy() { return maxEnergy; }
	/** 设置可存储的最大能量值 */
	public void setMaxEnergy(int max) { maxEnergy = max; }
	
	/** 能量接口 */
	private final IStorage storage = new IStorage() {
		
		@Override
		public boolean canReceive() {
			return isReceiveAllowable(null);
		}
		
		@Override
		public boolean canExtract() {
			return isExtractAllowable(null);
		}
		
		@Override
		public int getEnergyDemand() {
			return getDemandEnergy();
		}
		
		@Override
		public EleEnergy receiveEnergy(EleEnergy energy, boolean simulate) {
			if (world.isRemote || !canReceive()) return EMPTY_ENERGY.copy();
			EleEnergy real = energy.min(energy.copy(getMaxEnergy() - nowEnergy));
			if (simulate) return real;
			if (!onReceive(real)) return EMPTY_ENERGY.copy();
			nowEnergy += real.getCapacity();
			return real;
		}
		
		@Override
		public EleEnergy extractEnergy(int energy, boolean simulate) {
			if (world.isRemote) return EMPTY_ENERGY.copy();
			if (canExtract()) {
				//若存储能量或需要输出的能量小于等于0则直接返回
				if (nowEnergy <= 0 || energy <= 0) return EMPTY_ENERGY.copy();
				
				//计算应该输出的电能
				int real = Math.min(nowEnergy, energy);
				//若需要输出的电压不在可输出的范围则输出最适电压
				EleEnergy reEnergy = new EleEnergy(real, getExVoltage());
				if (!simulate) {
					if (!onExtract(reEnergy)) return EMPTY_ENERGY.copy();
					nowEnergy -= real;
				}
				return reEnergy;
			}
			return EMPTY_ENERGY.copy();
		}
		
		@Override
		public boolean isReAllowable(EnumFacing facing) {
			return EleTileEntity.this.isReceiveAllowable(facing);
		}
		
		@Override
		public boolean isExAllowable(EnumFacing facing) {
			return EleTileEntity.this.isExtractAllowable(facing);
		}
		
		@Override
		public boolean canLink(EnumFacing facing) {
			return EleTileEntity.this.canLinkEle(facing);
		}
		@Override
		public boolean link(BlockPos pos) {
			return EleTileEntity.this.linkEle(pos);
		}
		@Override
		public boolean unLink(BlockPos pos) { return EleTileEntity.this.unLink(pos); }
		@Override
		public boolean isLink(BlockPos pos) {
			return linkedBlocks.contains(pos);
		}
		
	};
	
}

File: src/main/java/top/kmar/mi/api/electricity/EleTileEntity.kt
package top.kmar.mi.api.electricity

import net.minecraft.util.EnumFacing
import net.minecraftforge.common.capabilities.Capability
import top.kmar.mi.api.araw.interfaces.AutoSave
import top.kmar.mi.api.electricity.cables.EleCableEntity
import top.kmar.mi.api.electricity.caps.ElectricityCapability.capObj
import top.kmar.mi.api.electricity.caps.EmptyElectricityCap
import top.kmar.mi.api.electricity.caps.IElectricityCap
import top.kmar.mi.api.tools.BaseTileEntity
import top.kmar.mi.api.utils.container.IndexEnumMap
import top.kmar.mi.api.utils.expands.computeIfAbsent
import java.util.function.Consumer

/**
 * 电力方块的父级TE
 * @author EmptyDreams
 */
abstract class EleTileEntity : BaseTileEntity() {

    /** 存储方块六个方向上的连接数据 */
    @field:AutoSave
    protected val linkData = IndexEnumMap(EnumFacing.values())
    /** 缓存各个方向上的`cap` */
    private val capArray = Array<IElectricityCap?>(EnumFacing.values().size) { null }

    override fun hasCapability(capability: Capability<*>, facing: EnumFacing?): Boolean {
        if (capability == capObj) {
            if (facing != null) return linkData[facing]
            for (value in EnumFacing.values()) {
                if (linkData[value]) return true
            }
            return false
        }
        return super.hasCapability(capability, facing)
    }

    override fun <T : Any?> getCapability(capability: Capability<T>, facing: EnumFacing?): T? {
        if (capability == capObj) return capObj.cast(getEleCap(facing))
        return super.getCapability(capability, facing)
    }

    /**
     * 检查输入的能量
     * @param input 输入的能量
     * @param minVoltage 最低电压
     * @param maxVoltage 最高电压
     * @param under 欠压时触发
     * @param over 过压时触发
     * @return 是否可以正常工作
     */
    protected inline fun checkEnergy(
        input: EleEnergy, minVoltage: Int, maxVoltage: Int,
        empty: () -> Unit, under: () -> Unit, over: () -> Unit
    ): Boolean {
        if (input.isEmpty) empty()
        else if (input.voltage > maxVoltage) over()
        else if (input.voltage < minVoltage) under()
        else return true
        return false
    }

    /**
     * 构建指定方向上的`cap`，
     * @param facing 指定的方向，该方向一定已经连接方块
     * @return 构建的`cap`对象，值会被缓存
     */
    protected open fun buildCap(facing: EnumFacing): IElectricityCap {
        return EmptyElectricityCap
    }

    /** 使指定方向上的`cap`缓存失效 */
    protected fun invalidCap(facing: EnumFacing) {
        capArray[facing.index] = null
    }

    /**
     * 获取指定方向上的[IElectricityCap]
     * @param facing 指定方向，如果为`null`则按照[EnumFacing]中的下标顺序查找第一个`cap`
     * @return 查找到的`cap`，如果没有找到则返回`null`
     */
    fun getEleCap(facing: EnumFacing?): IElectricityCap? {
        if (facing == null) {
            for (value in EnumFacing.values()) {
                return getEleCap(value) ?: continue
            }
            return null
        }
        if (!linkData[facing]) return null
        return capArray.computeIfAbsent(facing.index) { buildCap(facing) }
    }

    /** 向周围网络请求能量 */
    open fun requestEnergy(maxEnergy: Int): EleEnergy {
        var energy = maxEnergy
        var result = EleEnergy.empty
        eachLinked {
            val cable = world.getTileEntity(pos.offset(it)) as? EleCableEntity ?: return@eachLinked true
            val request = cable.requestEnergy(energy)
            energy -= request.capacity
            result = result.merge(request)
            energy != 0
        }
        return result
    }

    /** 统计连接了几个方向 */
    open fun countLinks(): Int {
        var result = 0
        for (i in capArray.indices) {
            if (linkData[EnumFacing.values()[i]]) ++result
        }
        return result
    }

    /** 判断指定方向是否连接 */
    open fun isLink(facing: EnumFacing): Boolean = linkData[facing]

    /**
     * 连接指定方向
     * @return 是否连接成功，重复连接返回`true`
     */
    open fun link(facing: EnumFacing): Boolean {
        linkData[facing] = true
        return true
    }

    /** 移除指定方向的连接 */
    open fun unlink(facing: EnumFacing) {
        linkData[facing] = false
    }

    /**
     * 遍历连接的方向
     * @param breakConsumer 遍历任务，返回`false`表示终止遍历
     */
    fun eachLinked(breakConsumer: (EnumFacing) -> Boolean) {
        for (value in EnumFacing.values()) {
            if (isLink(value)) {
                if (!breakConsumer(value)) break
            }
        }
    }

    /** 遍历所有连接的方向 */
    fun eachAllLinked(consumer: Consumer<EnumFacing>) {
        for (value in EnumFacing.values()) {
            if (isLink(value)) {
                consumer.accept(value)
            }
        }
    }

    /** 产生一个爆炸 */
    protected fun explode(strength: Float, isFlaming: Boolean) {
        world.setBlockToAir(pos)
        world.newExplosion(
            null, pos.x + 0.5, pos.y + 0.5, pos.z + 0.5, strength, isFlaming, true
        )
    }

}